### ISSUE TITLE
Support equality checking on arbitrary types

### ISSUE THREAD

**Issue by cgranade** [microsoft/qsharp-compiler#203]
**Is your feature request related to a problem? Please describe.**
At the moment, the `==` operator is not supported for values whose type is given by a type parameter (see screenshot from an IQ# session below). Since all types in Q# are value types, however, we should be able to support `==` between arbitrary types, including type parameters.

**Additional context**
![image](https://user-images.githubusercontent.com/31516/66951070-74206800-f00e-11e9-8f26-eb57eacea01f.png)

**Comment by bettinaheim** [comment]
Here is roughly the list of things that would need to be done: 
- Qubit equality comparison in compiler and runtime
- Array equality comparison in compiler and QArray
- Tuple equality comparison in compiler and QTuple
- UDT equality comparison in compiler and [runtime (UDT class) or code gen]
- Basic callable equality comparison in compiler and runtime
- Partial application equality comparison in runtime
Let me know if you are up for picking some of that up.

**Comment by samarsha** [comment]
Since there are good reasons to want to prevent equality on callable types, but a generic type parameter `'T` could potentially be instantiated as a callable type, it looks like this is blocked by #557.

**Comment by samarsha** [comment]
However, we could add equality comparisons for all of the other *concrete* types relatively easily (excludes callables and type parameters)... maybe that should be a new issue.

**Comment by alan-geller** [comment]
I'm going to transfer this issue to the qsharp-language repo, since that's now the proper place for language suggestions.
